diff --git a/queue.c b/queue.c
index 7725d0d..d1b0cff 100644
--- a/queue.c
+++ b/queue.c
@@ -2092,7 +2092,7 @@ _dispatch_mgr_root_queue_init(void)
 	(void)dispatch_assume_zero(pthread_attr_setdetachstate(attr,
 			PTHREAD_CREATE_DETACHED));
 #if !DISPATCH_DEBUG
-	(void)dispatch_assume_zero(pthread_attr_setstacksize(attr, 64 * 1024));
+	(void)dispatch_assume_zero(pthread_attr_setstacksize(attr, MAX(PTHREAD_STACK_MIN, 64 * 1024)));
 #endif
 #if HAVE_PTHREAD_WORKQUEUE_QOS
 	qos_class_t qos = _dispatch_mgr_sched.qos;
